fsspec/asyn.py

```diff
@@ -378,18 +378,19 @@ async def _put(
             lpath = make_path_posix(lpath)
         fs = LocalFileSystem()
         lpaths = fs.expand_path(lpath, recursive=recursive)
-        dirs = [l for l in lpaths if os.path.isdir(l)]
-        rdirs = other_paths(dirs, rpath)
+        rpaths = other_paths(lpaths, rpath)
+
+        rdirs = [r for l, r in zip(lpaths, rpaths) if os.path.isdir(l)]
+        file_pairs = [(l, r) for l, r in zip(lpaths, rpaths) if not os.path.isdir(l)]
```

Comment:
Could we possibly avoid calling `isdir` on every file twice?

Reply:
This is now called just once per file